File: ruby/hecks/behaviors/aggregate_state.rb
# Hecks::Behaviors::AggregateState
#
# Per-instance aggregate state held by the in-memory runtime.
# Mirrors hecks_life/src/runtime/mod.rs `AggregateState`. Pure
# field bag; the Interpreter and dispatch loop apply mutations
# and read fields.
#
#   state = AggregateState.new("1")
#   state.set("status", Value.from("approved"))
#   state.get("status").to_display     # => "approved"
require_relative "value"

module Hecks
  module Behaviors
    class AggregateState
      attr_reader :id, :fields

      def initialize(id)
        @id = id.to_s
        @fields = {}
      end

      def get(field)
        @fields[field.to_s] || Value.null
      end

      def set(field, value)
        @fields[field.to_s] = value.is_a?(Value) ? value : Value.from(value)
      end

      def append(field, value)
        current = @fields[field.to_s]
        items = current && current.list? ? current.raw.dup : []
        items << (value.is_a?(Value) ? value : Value.from(value))
        @fields[field.to_s] = Value.new(:list, items)
      end

      def increment(field, amount = 1)
        n = (get(field).numeric || 0) + amount
        @fields[field.to_s] = Value.from(n.to_i)
      end

      def decrement(field, amount = 1)
        n = (get(field).numeric || 0) - amount
        @fields[field.to_s] = Value.from(n.to_i)
      end

      # Float-aware variant for fractional increments like 0.01.
      # Stores the result as a Str (matching Rust runtime's behavior)
      # so it round-trips through the DSL's numeric coercion.
      def increment_float(field, amount)
        n = (get(field).numeric || 0.0) + amount
        @fields[field.to_s] = Value.new(:str, format_numeric(n))
      end

      def format_numeric(n)
        n == n.to_i ? n.to_i.to_s : n.to_s
      end

      def toggle(field)
        current = get(field)
        new_val = current.kind == :bool ? !current.raw : true
        @fields[field.to_s] = Value.from(new_val)
      end
    end
  end
end

File: rust/src/runtime/aggregate_state.rs
//! AggregateState — dynamic bag of fields
//!
//! Aggregates are not typed structs — they're Value maps
//! shaped by the Bluebook IR at runtime.
//!
//! Usage:
//!   let mut state = AggregateState::new("pizza_1");
//!   state.set("name", Value::Str("Margherita".into()));
//!
//! [antibody-exempt: i106 dsl-mutation-primitives — adds `set_float`
//!  for Multiply / Clamp / Decay. Same retirement contract as ir.rs.]

use super::Value;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct AggregateState {
    pub id: String,
    pub fields: HashMap<String, Value>,
    /// Set true by a `then_delete` mutation. The command dispatcher
    /// checks this after `apply_mutations` and calls
    /// `Repository::delete` instead of `Repository::save` when
    /// present. Default false ; only retire-style commands flip it.
    pub deleted: bool,
}

impl AggregateState {
    pub fn new(id: &str) -> Self {
        AggregateState {
            id: id.to_string(),
            fields: HashMap::new(),
            deleted: false,
        }
    }

    pub fn get(&self, field: &str) -> &Value {
        self.fields.get(field).unwrap_or(&Value::Null)
    }

    pub fn set(&mut self, field: &str, value: Value) {
        self.fields.insert(field.to_string(), value);
    }

    pub fn append(&mut self, field: &str, value: Value) {
        let list = self
            .fields
            .entry(field.to_string())
            .or_insert_with(|| Value::List(vec![]));
        if let Value::List(ref mut v) = list {
            v.push(value);
        }
    }

    pub fn increment(&mut self, field: &str, amount: i64) {
        let current = match self.fields.get(field) {
            Some(Value::Int(n)) => *n,
            _ => 0,
        };
        self.fields
            .insert(field.to_string(), Value::Int(current + amount));
    }

    pub fn decrement(&mut self, field: &str, amount: i64) {
        let current = match self.fields.get(field) {
            Some(Value::Int(n)) => *n,
            _ => 0,
        };
        self.fields
            .insert(field.to_string(), Value::Int(current - amount));
    }

    /// Float-aware increment. The Bluebook DSL allows fractional
    /// increments (`then_set :fatigue, increment: 0.01`); the int-only
    /// `increment()` would silently round 0.01 to 1 (after the
    /// `unwrap_or(1)` fallback in the interpreter), so fatigue would
    /// track beats one-to-one instead of accumulating slowly. This
    /// path stores the result as a `Str` representation of the float
    /// so it round-trips through the DSL's numeric coercion.
    pub fn increment_float(&mut self, field: &str, amount: f64) {
        let current = current_numeric(self.fields.get(field));
        let new_val = current + amount;
        self.fields
            .insert(field.to_string(), Value::Str(format_numeric(new_val)));
    }

    pub fn decrement_float(&mut self, field: &str, amount: f64) {
        let current = current_numeric(self.fields.get(field));
        let new_val = current - amount;
        self.fields
            .insert(field.to_string(), Value::Str(format_numeric(new_val)));
    }

    /// Float-aware set — i106 Multiply / Decay / Clamp. The new value
    /// is rendered as the same numeric Str representation
    /// `increment_float` uses, so downstream comparisons (givens,
    /// queries) coerce identically across +/- and ×/clamp paths.
    pub fn set_float(&mut self, field: &str, value: f64) {
        self.fields
            .insert(field.to_string(), Value::Str(format_numeric(value)));
    }

    pub fn toggle(&mut self, field: &str) {
        let current = match self.fields.get(field) {
            Some(Value::Bool(b)) => *b,
            _ => false,
        };
        self.fields
            .insert(field.to_string(), Value::Bool(!current));
    }
}

fn current_numeric(v: Option<&Value>) -> f64 {
    match v {
        Some(Value::Int(n)) => *n as f64,
        Some(Value::Str(s)) => s.parse::<f64>().unwrap_or(0.0),
        _ => 0.0,
    }
}

/// Format a numeric value for storage. Whole-valued floats stringify
/// as "3" (not "3.0") so they keep parity with Int(3) for downstream
/// equality comparisons; fractional values keep their decimal form.
fn format_numeric(n: f64) -> String {
    if n == n.trunc() && n.abs() < 1e15 {
        format!("{}", n as i64)
    } else {
        format!("{}", n)
    }
}
